File: sdk/go/cytindexer/tool_entries.go
package cytindexer

import "unicode/utf8"

func approximateTokenCount(text string) int {
	count := 0
	for _, c := range text {
		if c < 128 {
			count++
		} else {
			count += 2
		}
	}
	return count/2 + 1
}

// TruncateDescription truncates text to at most maxTokens (approximate).
func TruncateDescription(description string, maxTokens int) string {
	if description == "" {
		return ""
	}
	if approximateTokenCount(description) <= maxTokens {
		return description
	}
	suffix := "..."
	suffixTokens := approximateTokenCount(suffix)
	bodyBudget := maxTokens - suffixTokens
	if bodyBudget <= 0 {
		return suffix
	}

	runes := []rune(description)
	lo, hi := 0, len(runes)
	for lo < hi {
		mid := (lo + hi + 1) / 2
		if approximateTokenCount(string(runes[:mid])) <= bodyBudget {
			lo = mid
		} else {
			hi = mid - 1
		}
	}
	body := string(runes[:lo])
	if sp := lastSpace(body); sp > 0 {
		body = body[:sp]
	}
	return body + suffix
}

func lastSpace(s string) int {
	for i := len(s) - 1; i >= 0; i-- {
		if s[i] == ' ' {
			return i
		}
		if s[i] >= utf8.RuneSelf {
			r, size := utf8.DecodeLastRuneInString(s[:i+1])
			if r == ' ' {
				return i - size + 1
			}
		}
	}
	return -1
}

func anthropicInputSchema(tool map[string]any) any {
	if s, ok := tool["input_schema"]; ok {
		return s
	}
	if s, ok := tool["inputSchema"]; ok {
		return s
	}
	if s, ok := tool["parameters"]; ok {
		return s
	}
	return map[string]any{}
}

// IsCatalogToolEntry reports whether tool matches catalog entry shape.
func IsCatalogToolEntry(tool any) bool {
	obj, ok := AsObject(tool)
	if !ok {
		return false
	}
	id, ok := obj["id"].(string)
	_, hasSchema := AsObject(obj["full_schema"])
	return ok && id != "" && hasSchema
}

// PrepareToolEntry builds one catalog entry from tool metadata.
func PrepareToolEntry(serverName, name, description string, inputSchema any) map[string]any {
	fullSchema := map[string]any{
		"id": name, "name": name, "description": description, "inputSchema": inputSchema,
	}
	return map[string]any{
		"id": name, "server": serverName, "tool": name,
		"summary": TruncateDescription(description, 60), "full_schema": fullSchema,
	}
}

// AnthropicToolToCatalogEntry converts one Anthropic API tool to a catalog entry.
func AnthropicToolToCatalogEntry(tool any) (map[string]any, bool) {
	obj, ok := AsObject(tool)
	if !ok {
		return nil, false
	}
	name, ok := obj["name"].(string)
	if !ok || name == "" {
		return nil, false
	}
	description := StrField(obj, "description")
	entry := PrepareToolEntry("", name, description, anthropicInputSchema(obj))
	return entry, true
}

// NormalizeToolsForCatalog normalizes a tool list for indexing.
func NormalizeToolsForCatalog(tools []any) ([]any, []any) {
	entries := make([]any, 0, len(tools))
	var allEnums []any
	for _, tool := range tools {
		var entry any
		if IsCatalogToolEntry(tool) {
			entry = CloneValue(tool)
		} else if e, ok := AnthropicToolToCatalogEntry(tool); ok {
			entry = e
		} else {
			continue
		}
		if obj, ok := AsObject(entry); ok {
			if fs, ok := AsObject(obj["full_schema"]); ok {
				if schema, ok := fs["inputSchema"]; ok {
					allEnums = append(allEnums, CollectEnums(schema)...)
				}
			}
		}
		entries = append(entries, entry)
	}
	return entries, allEnums
}

// BuildCatalogFromTools builds a decomposed catalog index from API tools or catalog entries.
func BuildCatalogFromTools(tools []any) CatalogIndex {
	entries, enums := NormalizeToolsForCatalog(tools)
	return BuildCatalogIndex(entries, enums)
}

// AnthropicToolsToCatalogEntries converts Anthropic API tools to catalog entries.
func AnthropicToolsToCatalogEntries(tools []any) ([]any, []any) {
	var entries []any
	var allEnums []any
	for _, tool := range tools {
		entry, ok := AnthropicToolToCatalogEntry(tool)
		if !ok {
			continue
		}
		if fs, ok := AsObject(entry["full_schema"]); ok {
			if schema, ok := fs["inputSchema"]; ok {
				allEnums = append(allEnums, CollectEnums(schema)...)
			}
		}
		entries = append(entries, entry)
	}
	return entries, allEnums
}

File: sdk/rust/cyt-indexer/src/tool_entries.rs
//! Convert Anthropic API tools into catalog entries — tiktoken-backed description truncation.

use crate::build::{CatalogIndex, build_catalog_index};
use crate::paths::collect_enums;
use serde_json::{Value, json};

use crate::tiktoken;

/// Truncate text to at most `max_tokens` (tiktoken), preferring a word boundary.
#[must_use]
pub fn truncate_description(description: &str, max_tokens: usize) -> String {
    tiktoken::truncate_description_or_passthrough(description, max_tokens)
}

fn anthropic_input_schema(tool: &Value) -> Value {
    tool.get("input_schema")
        .or_else(|| tool.get("inputSchema"))
        .or_else(|| tool.get("parameters"))
        .cloned()
        .unwrap_or_else(|| json!({}))
}

/// True when `tool` already matches catalog entry shape (`id` + `full_schema`).
pub fn is_catalog_tool_entry(tool: &Value) -> bool {
    tool.get("id")
        .and_then(Value::as_str)
        .is_some_and(|id| !id.is_empty())
        && tool.get("full_schema").is_some_and(Value::is_object)
}

/// Build one catalog entry from tool metadata (no file I/O).
#[must_use]
pub fn prepare_tool_entry(
    server_name: &str,
    name: &str,
    description: &str,
    input_schema: &Value,
) -> Value {
    let full_schema = json!({
        "id": name,
        "name": name,
        "description": description,
        "inputSchema": input_schema,
    });
    json!({
        "id": name,
        "server": server_name,
        "tool": name,
        "summary": truncate_description(description, 60),
        "full_schema": full_schema,
    })
}

/// Convert one Anthropic `{ name, description, input_schema }` tool to a catalog entry.
pub fn anthropic_tool_to_catalog_entry(tool: &Value) -> Option<Value> {
    let name = tool.get("name").and_then(Value::as_str)?;
    if name.is_empty() {
        return None;
    }
    let description = tool
        .get("description")
        .and_then(Value::as_str)
        .unwrap_or("");
    let input_schema = anthropic_input_schema(tool);
    Some(prepare_tool_entry("", name, description, &input_schema))
}

/// Normalize a tool list (Anthropic API and/or catalog entries) for indexing.
#[must_use]
pub fn normalize_tools_for_catalog(tools: &[Value]) -> (Vec<Value>, Vec<Value>) {
    let mut entries = Vec::with_capacity(tools.len());
    let mut all_enums = Vec::new();

    for tool in tools {
        let entry = if is_catalog_tool_entry(tool) {
            tool.clone()
        } else {
            match anthropic_tool_to_catalog_entry(tool) {
                Some(entry) => entry,
                None => continue,
            }
        };
        if let Some(schema) = entry.pointer("/full_schema/inputSchema") {
            all_enums.extend(collect_enums(schema));
        }
        entries.push(entry);
    }

    (entries, all_enums)
}

/// Build a decomposed catalog index from Anthropic API tools or pre-built catalog entries.
#[must_use]
pub fn build_catalog_from_tools(tools: &[Value]) -> CatalogIndex {
    let (entries, enums) = normalize_tools_for_catalog(tools);
    build_catalog_index(&entries, &enums)
}

/// Convert Anthropic API tools to catalog entries and collected enum values.
#[must_use]
pub fn anthropic_tools_to_catalog_entries(tools: &[Value]) -> (Vec<Value>, Vec<Value>) {
    let mut entries = Vec::new();
    let mut all_enums = Vec::new();
    for tool in tools {
        let Some(entry) = anthropic_tool_to_catalog_entry(tool) else {
            continue;
        };
        if let Some(schema) = entry.pointer("/full_schema/inputSchema") {
            all_enums.extend(collect_enums(schema));
        }
        entries.push(entry);
    }
    (entries, all_enums)
}
